src/dataflow/src/source/mod.rs

```diff
@@ -849,6 +853,221 @@ enum MessageProcessing {
+            sender,
+            tick_duration,
+        }
+    }
+
+    /// Start a transaction at a particular point in time. The timestamper will freeze its internal
+    /// clock while a transaction is active.
+    pub async fn start_tx<'a>(&'a self) -> SourceTransaction<'a> {
+        SourceTransaction {
+            timestamp: self.inner.read().await,
+            sender: &self.sender,
+        }
+    }
+
+    /// Record an insertion of a row
+    pub async fn insert(&self, row: Row) -> anyhow::Result<()> {
+        self.start_tx().await.insert(row).await
+    }
+
+    /// Record a deletion of a row
+    pub async fn delete(&self, row: Row) -> anyhow::Result<()> {
+        self.start_tx().await.delete(row).await
+    }
+
+    /// Records an error. After this method is called the source will permanently be in an errored
+    /// state
+    pub async fn error(&self, err: SourceError) -> anyhow::Result<()> {
+        let timestamp = self.inner.read().await;
+        self.sender
+            .send(Event::Message(*timestamp, Err(err)))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+
+    /// Attempts to monotonically increase the current timestamp and provides a Progress message to
+    /// timely.
+    ///
+    /// This method will wait for all current transactions to commit before advancing the clock and
+    /// will cause any new requests for transactions to wait for the tick to complete before
+    /// starting.  This is due to the write-preferring behaviour of the tokio RwLock.
+    async fn tick(&self) -> anyhow::Result<()> {
+        tokio::time::sleep(self.tick_duration).await;
+        let mut timestamp = self.inner.write().await;
+        let now = UNIX_EPOCH
+            .elapsed()
+            .expect("system clock before 1970")
+            .as_millis()
+            .try_into()
+            .expect("materialize has existed for more than 500M years");
+        if *timestamp < now {
+            *timestamp = now;
+            self.sender
+                .send(Event::Progress(Some(*timestamp)))
+                .await
+                .or_else(|_| Err(anyhow!("channel closed")))?;
+        }
+        Ok(())
+    }
+}
+
+/// Simple sources must implement this trait. Sources will then get created as part of the
+/// [`create_source_simple`] function.
+///
+/// Each simple source is given access to a timestamper instance that can be used to insert or
+/// retract rows for this source. See the API of the [Timestamper](Timestamper) for more details.
+#[async_trait]
+pub trait SimpleSource {
+    /// Consumes the instance of this SimpleSource and converts it into an async state machine that
+    /// submits rows using the provided timestamper.
+    ///
+    /// Implementors should return an Err(_) if an unrecoverable error is encountered or Ok(()) when
+    /// they have finished consuming the upstream data.
+    async fn start(self, timestamper: &Timestamper) -> Result<(), SourceError>;
```

Comment:
so if I'm understanding this right - implementers can use the `Timestamper` `insert` / `delete` messages and they could also use `error` to send error messages to the source without shutting down the timestamper, and they can also return an `Err` to cause everything to shut down is that accurate? specifically it seems like theres two ways implementers can send error information and I want to make sure a) thats an accurate understanding and b) if so, could we get some guidance on when to use which?

Reply:
good catch! The `error` method was supposed to be private, I'll fix that